### PR TITLE
Solution for 'Make animation parameter in [AnimatedList.removeItem] o…

### PR DESCRIPTION
**Pull request by hyperpanthera**
…ptional' #29844

https://github.com/flutter/flutter/issues/29844

## Description

*It would be great if the animation parameter in [AnimatedList.deleteInsert] could be made optional, simply so we can use the [Dismissible] widget along with an [AnimatedList] without the animations overlapping each other when an item is removed by [Dismissible] and [AnimatedList.deleteInsert] is required, so no logical errors within the [AnimatedList] occurs.*

## Related Issues

*#29844*

## Tests

I added the following tests:

*Replace this with a list of the tests that you added as part of this PR. A change in behaviour with no test covering it
will likely get reverted accidentally sooner or later. PRs must include tests for all changed/updated/fixed behaviors. See [Test Coverage].*

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by hyperpanthera** [comment]
I signed it!

> Thanks for your pull request. It looks like this may be your first contribution to a Google open source project (if not, look below for help). Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
> 📝 **Please visit https://cla.developers.google.com/ to sign.**
> 
> Once you've signed (or fixed any issues), please reply here (e.g. `I signed it!`) and we'll verify it.
> 
> #### What to do if you already signed the CLA
> ##### Individual signers
> * It's possible we don't have your GitHub username or you're using a different email address on your commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> 
> ##### Corporate signers
> * Your company has a Point of Contact who decides which employees are authorized to participate. Ask your POC to be added to the group of authorized contributors. If you don't know who your Point of Contact is, direct the Google project maintainer to [go/cla#troubleshoot](http://go/cla#troubleshoot) ([Public version](https://opensource.google.com/docs/cla/#troubleshoot)).
> * The email used to register you as an authorized contributor must be the email used for the Git commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * The email used to register you as an authorized contributor must also be [attached to your GitHub account](https://github.com/settings/emails).
> 
> ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Fflutter%2Fflutter%2Fpull%2F32501) for more info**.

I signed it!

**Comment by goderbauer** [comment]
I do not understand what the motivation behind this change is. Can you provide more context?

Also: There are multiple checks failing. Please fix those issues. And you will also have to add tests for any new functionality that you're adding to ensure that we do not break this in the future.

**Comment by hyperpanthera** [comment]
@goderbauer Imagine you want to use Dismissible combined with an AnimatedList. Now let's say you have a button to normally delete AnimatedList. Okay no problem. But what if you can delete it also by dismissing? Then you still need to delete the entry in AnimatedList with .delete. But since .delete requires an animation we cannot do that unless we make the animation parameter optional...

**Comment by goderbauer** [comment]
What is `.deleteInsert`?

So you just want to delete an item from an AnimatedList without animation? Can you just pass an animation duration of Zero to removeItem? Or alternatively, when you remove the item you just modify the `itemBuilder` passed to the AnimatedList to no longer build that particular item>

**Comment by hyperpanthera** [comment]
Sorry I meant removeInsert.

And those type of solutions are not really clean and hacky and resulted somewhat to problems in my case.

**Comment by goderbauer** [comment]
What are the problems that you saw with those approaches?

**Comment by hyperpanthera** [comment]
@goderbauer It gave an error for a second. I think it is due to the fact that the deleteInsert function tries to call .reverse on the Animation Controller. Thus if no proper animation is specified you get an error.

**Comment by goderbauer** [comment]
Let's continue the discussion over on the issue #29844, where I have posted a code example that seems to work just fine with setting the duration to zero and using a Dismissible.

**Review by Piinks — COMMENTED** [review]
Hi @hyperpanthera , did @goderbauer's solution in #29844 help with the issue you have been experiencing?
Would you like to work further on this PR?

**Comment by hyperpanthera** [comment]
@Piinks It seems to work so far in my tests but I think it is a hacky solution because we still pass an animation that we do not use (giving it the attribute Duration.zero). I still vote to make the animation optional. Why force an animation?

**Comment by hyperpanthera** [comment]
@Piinks @goderbauer Please review the new code I pulled. It is no breaking change anymore. It allows that 'null' can be passed as animation parameter. In this case, no animation will be rendered. If it is committed, users will not require to change their code while still having the option to not pass an animation at all, thus making the widget more flexible and less restrictive.

**Comment by hyperpanthera** [comment]
All checks passed (**_flutter-build_** is currentl broken).

File: packages/flutter/lib/src/widgets/animated_list.dart [hunk 1]
```diff
@@ -333,7 +332,31 @@ class AnimatedListState extends State<AnimatedList> with TickerProviderStateMixi
...
+          if (item.itemIndex > outgoingItem.itemIndex)
+            item.itemIndex -= 1;
+        }
+
+        setState(() {
+          _itemsCount -= 1;
+        });
+      });
+    }
+    /// [5/24/2019]
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/animated_list.dart]: What is this date breadcrumb here for? What does it link to?

Comment (hyperpanthera) [reply packages/flutter/lib/src/widgets/animated_list.dart]: @Piinks

The date and part I edited.

You can remove the comment.

Sorry.

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/animated_list.dart]: Can you remove this comment?

**Review by goderbauer — COMMENTED** [review]
We should probably check if the duration is zero and if that's true skip all the work related to the animation. We should also update the documentation to reflect this behavior.

This will also need tests to ensure that we never regress/break this new behavior.

File: packages/flutter/lib/src/widgets/animated_list.dart [hunk 2]
```diff
@@ -314,9 +314,8 @@ class AnimatedListState extends State<AnimatedList> with TickerProviderStateMixi
   /// This method's semantics are the same as Dart's [List.remove] method:
   /// it decreases the length of the list by one and shifts all items at or
   /// before [index] towards the beginning of the list.
-  void removeItem(int index, AnimatedListRemovedItemBuilder builder, { Duration duration = _kDuration }) {
+  void removeItem(int index, AnimatedListRemovedItemBuilder builder, {Duration duration = _kDuration}) {
```
[the remark this answers is not in the document]
Comment (hyperpanthera) [reply packages/flutter/lib/src/widgets/animated_list.dart]: @goderbauer Done too!

**Comment by hyperpanthera** [comment]
> We should probably check if the duration is zero and if that's true skip all the work related to the animation. We should also update the documentation to reflect this behavior.
> 
> This will also need tests to ensure that we never regress/break this new behavior.

Please review the new code I pulled.

I added both cases now where one can pass either 'null' as animation or [Duration.zero] to skip the animation work, I think to give the possibility for both is good, otherwise one would still need to define an animation which is boilerplate code.

File: packages/flutter/lib/src/widgets/animated_list.dart (hunk 1 above)
Comment (hyperpanthera) [reply packages/flutter/lib/src/widgets/animated_list.dart]: @goderbauer Done.

**Review by Piinks — COMMENTED** [review]
Hi @hyperpanthera, thank you for your continued work on this. We still need tests for this new behavior.

**Comment by hyperpanthera** [comment]
Hi @Piinks, how can I provide these tests or are they run by your team?

**Comment by Piinks** [comment]
It looks like for this change the appropriate place to write a test would be here: `packages/flutter/test/widgets/animated_list_test.dart`
Writing a test to validate the expected new behavior helps to make sure we do not regress or break this change later on. 
Here is some information on writing widget tests:
https://flutter.dev/docs/cookbook/testing/widget/introduction

File: packages/flutter/lib/src/widgets/animated_list.dart [hunk 3]
```diff
@@ -316,7 +316,6 @@ class AnimatedListState extends State<AnimatedList> with TickerProviderStateMixi
   /// before [index] towards the beginning of the list.
   void removeItem(int index, AnimatedListRemovedItemBuilder builder, { Duration duration = _kDuration }) {
     assert(index != null && index >= 0);
-    assert(builder != null);
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/animated_list.dart]: this assertion should be modified to express that builder is only allowed to be null when duration is zero.

File: packages/flutter/lib/src/widgets/animated_list.dart [hunk 4]
```diff
@@ -333,7 +332,32 @@ class AnimatedListState extends State<AnimatedList> with TickerProviderStateMixi
         ..sort();
     });
 
-    controller.reverse().then<void>((void value) {
+    if (builder != null && duration != Duration.zero) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/animated_list.dart]: When duration is zero it doesn't matter what value builder has.

File: packages/flutter/lib/src/widgets/animated_list.dart [hunk 5]
```diff
@@ -333,7 +332,32 @@ class AnimatedListState extends State<AnimatedList> with TickerProviderStateMixi
...
+      controller.reverse().then<void>((void value) {
+        _removeActiveItemAt(_outgoingItems, outgoingItem.itemIndex).controller.dispose();
+
+        // Decrement the incoming and outgoing item indices to account
+        // for the removal.
+        for (_ActiveItem item in _incomingItems) {
+          if (item.itemIndex > outgoingItem.itemIndex)
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/animated_list.dart]: We should avoid the duplication between the two branches.

File: packages/flutter/lib/src/widgets/animated_list.dart [hunk 6]
```diff
@@ -333,7 +332,32 @@ class AnimatedListState extends State<AnimatedList> with TickerProviderStateMixi
...
+            item.itemIndex -= 1;
+        }
+
+        setState(() {
+          _itemsCount -= 1;
+        });
+      });
+    }
+    /// In cases when the playback of an animation is not desired.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/animated_list.dart]: Can you remove this comment? It doesn't really help in understanding the code. This should be in the PR description.

**Review by goderbauer — COMMENTED** [review]
This will also need a test. Can you please add one?

**Comment by goderbauer** [comment]
@hyperpanthera Are you still planning on coming back to this PR?

**Comment by hyperpanthera** [comment]
@goderbauer I don't know how this test thing works. I offered just a solution here that I also use on production without any error.

**Comment by goderbauer** [comment]
I'll close this for now until somebody from our team finds the time to add tests for this. Feel free to re-open if you find the time to update this PR with tests in the meantime.

**Pull request closed without merging**